[move-only] Teach SILGenApply how to emit subscripts with borrowed base values.

gottesmm opened the pull request:
I also added a bunch of tests that showed the behavior of subscripts/other accessors with the following combinations of semantics:

1. get only.
2. get/set.
3. get/modify.
4. read/set.
5. read/modify.

rdar://109746476

gottesmm committed:
[move-only] Teach SILGenApply how to emit subscripts with borrowed base values.

I also added a bunch of tests that showed the behavior of subscripts/other accessors with the following combinations of semantics:

1. get only.
2. get/set.
3. get/modify.
4. read/set.
5. read/modify.

rdar://109746476

gottesmm committed:
[move-only] When we emit a noncopyable temporary allocation, always attach a mark_must_check [consumable_and_assignable].

Most of the time SILGen already emits these correctly without having extra
copies, but in certain situations SILGen will emit copies that we need the move
checker to eliminate (e.x.: when we generate a yield). An additional benefit is
that this also will catch places where the frontend makes a mistake.

This also removes a bunch of "copy of noncopyable" types error that showed up in
the implicit compiler generated modify.

gottesmm commented:
@swift-ci smoke test

jckarter commented on test/SILGen/moveonly.swift:
```diff
@@ -1000,3 +1004,2094 @@ struct AddressOnlySetterTester : ~Copyable {
+public func testSubscriptGetOnly_BaseLoadable_ResultAddressOnly_Var() {
+    var m = LoadableSubscriptGetOnlyTester()
+    m = LoadableSubscriptGetOnlyTester()
+    m[0].nonMutatingFunc()
```
Since `m[0]` is implemented by a getter, the result is an independently consumable value. Should the formal access to `m` be shortened to end after `m[0]` returns? Does `let m0 = m[0]` work?

gottesmm replied:
It does. I added a test that validates that.

gottesmm committed:
[move-only] Make sure that we handle load [take] that are actually borrows correctly.

Before the previous commit, we didn't see load [take] very often since it occurs
mostly on temporaries where we treat the copy_addr as the relevant take. Now
that we are checking temporaries though, we need to support this behavior.

gottesmm committed:
[move-only] Restrict 508bf8ae2141f379d24dd0ae2902e5672fc52e9f so it only applies if a subscript has a _read accessor.

We want the result of getters to still be separate values.

gottesmm commented:
@swift-ci smoke test

jckarter commented on test/SILGen/moveonly.swift:
```diff
@@ -1000,3 +1004,2127 @@ struct AddressOnlySetterTester : ~Copyable {
+// CHECK: } // end sil function '$s8moveonly047testSubscriptGetOnly_BaseLoadable_ResultAddressE4_LetyyF'
+public func testSubscriptGetOnly_BaseLoadable_ResultAddressOnly_Let() {
+    let m = LoadableSubscriptGetOnlyTester()
+    m[0].nonMutatingFunc()
```
It looks like in this case the `nonMutatingFunc()` call still occurs inside of the borrow of `m`.

gottesmm replied:
This is not a result of this change since it is using a getter. I am pretty sure this is a result of us borrowing bases of accessors in an aggressive way which was part of the original implementation of accessors. IIRC @atrick ran into this when exploring non-escaping types.

jckarter commented on test/SILGen/moveonly.swift:
```diff
@@ -1000,3 +1004,2127 @@ struct AddressOnlySetterTester : ~Copyable {
+// CHECK: [[ACCESS:%.*]] = begin_access [read] [unknown] [[PROJECT]]
+// CHECK: [[MARK:%.*]] = mark_must_check [no_consume_or_assign] [[ACCESS]]
+// CHECK: [[LOAD:%.*]] = load_borrow [[MARK]]
+// Eventually, we need this end_apply to be around the nonMutatingFunc.
```
Since `AddressOnlyProtocol` is move-only, we can't copy it out of the `end_apply`/`end_borrow`/`end_access`, and the call to `nonMutatingMethod` does have to happen during the borrow. So "eventually" doesn't seem to cut it. Does the move only checker fix this up?

gottesmm replied:
It was an issue where SILGen was failing to pattern match. I did an update that fixes this.

gottesmm committed:
[move-only] Teach SILGen how to recognize a borrowed read subscript when the type also has a modify.

The form of the AST changes slightly when a type has a read and a modify.
Specifically, we now have a load on the subscript and an inout_expr on the base.
I dealt with this by making the inout_expr something that when we look for
storage we look through and by tweaking the load lookthrough code.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@jckarter I want to do some work this weekend that builds on top of this. I am happy to change anything else that you need changed post-commit.
